### DISCUSSION TITLE
冷门算法选讲

### DISCUSSION THREAD

**Discussion by 1205240810** [CSGrandeur/s-1problem1day1ac#510] [Algorithm]
[冷门算法选讲.pdf](https://github.com/CSGrandeur/s-1problem1day1ac/files/9794507/default.pdf)

# 冷门算法选讲
 
这个专题的目的还是从实用性出发，比赛马上就要到来，大家也都有自己的规划，这时候选一些复杂算法深入学习显然性价比不高，所以本专题主要从应用层面出发，介绍一些不需要太多训练就能掌握的套路性强的冷门算法。很多算法的证明过程也基本省略跳过处理。尽力做到只讲必要的原理和知道就能用的结论，还有一些拿来就能用的模板。

### 最基本的摩尔投票

摩尔投票法(Boyer–Moore majority vote algorithm)是一个比较冷门的算法，最初算法解决的问题是如何在任意多的候选人（选票无序），选出获得票数最多的那个。在算法竞赛中我们较多的用它来解决绝对众数的问题。

先从最基本情况出发，思考一个问题，如何找到n个数中的绝对众数，这里的绝对众数指的是这个数的数量占所有数的比例超过1/2。这个问题很直接的思路是可以直接排序做，但是时间复杂度并不优秀，并且如果要保留原序列，还需要额外的空间开销。不是一个很好的方式，这时候，我们就可以用摩尔投票的思想来解决这个问题。

首先说明一个事实，就是区间绝对众数只有1个数，摩尔投票算法是基于这个事实：每次从序列里选择两个不相同的数字删除掉（或称为“抵消”），最后剩下一个数字或几个相同的数字，就是出现次数大于总数一半的那个。

如果这样说很抽象，举个简单的例子帮助理解，又n个人打群架，大家都有自己的阵营，但是有一个阵营达成了垄断，就算其他所有阵容联合起来，也打不过这个最大的阵容。所以每次我们只要保证不打内战，每次让两个元素同归于尽，最后剩下的就是要找的这个元素。

实现起来也很容易，我们只需要记录一个元素now表示当前的元素，一个元素cnt记录当前元素now的数量。对于一张新元素，如果和now相同，cnt++，否则cnt--，如果cnt为0时候，新元素成为新的now。

```cpp
#include<bits/stdc++.h>
using namespace std;
using ll=long long;
int main()
{
    cin.tie(nullptr)->sync_with_stdio(false);
    ll now = 0, cnt = 0;
    for (int i=1;i<=n;i++)
    {
        if (cnt == 0)
            now = a[i];
        if (now == a[i])
            cnt++;
        else
            cnt--;
    }
    return 0;
}
```

注意这个算法只有在绝对众数存在的时候才能确保返回的答案正确性。、

这个算法是 O(n) 时间复杂度、 O(1) 空间复杂度的。

### 扩展摩尔投票

可以证明，出现次数超过  n/k  的数最多只有  k−1 个。否则必然违背数只有 n 个或者当前统计的是出现次数超过 n/k 的数的前提条件。

有了这个前提，我们还是按照上面的标准执行摩尔投票，在遍历数组过程中依次检查这k-1个候选数。执行即可

上述做法正确性的关键是：若存在出现次数超过 n/k 的数，最后必然会成为这 k−1 个候选者之一。

这个结论可以通过反证法证明，也可以直接感性理解一下，或者直接当结论处理，感兴趣可以阅读这篇博客了解更多[[摩尔投票（绝对众数） - OIer某罗 - 博客园 (cnblogs.com)](https://www.cnblogs.com/Zeardoe/p/16777614.html)](https://www.cnblogs.com/Zeardoe/p/16777614.html) 。

```CPP
#include<bits/stdc++.h>
using namespace std;
using ll=long long;
ll now[n],cnt[n];
int main()
{
    cin.tie(nullptr)->sync_with_stdio(false);
    for(int i=1;i<=n;i++)
    {
        int q = find(now,now+k,a[i])-now;
        if (q != k) { // 如果当前票投给了候选人之一
            cnt[q]++;
            continue;
        }
        int j = find(cnt,cnt+k,0)-cnt;
        if (j != k) { // 如果当前存在一个位置"虚位以待"
            now[j]=a[i];
            cnt[j] = 1;
            continue;
        }
        for(int j=0;j<k;j++)
        cnt[j]--;
    }
    return 0;
}
```

### 区间绝对众数

区间绝对众数问题也可以利用摩尔投票解决，这需要在线段树上进行摩尔投票。

首先摩尔投票其实是不符合结合律的，但是其实在这个维护区间众数的过程中，cnt并不关键，如果区间存在绝对众数，他是可以交换顺序记票的，所以仍然可以得到正确的绝对众数，只是需要额外验证。所以，我们可以在线段树上进行摩尔投票。在合并左右两个区间时，相当于是先对左右两边分别进行摩尔投票，再对两边抵消后的结果进行摩尔投票。

今年的noi 签到题就是这样一道利用了摩尔投票的数据结构题，一定程度上算是某种套路题，感兴趣可以尝试做做看。

### 康托展开

康托展开是一个全排列到一个自然数的双射，常用于构建hash表时的空间压缩。设有n个数(1,2,3,4,...,n) ，可以有组成不同(n!种)的排列组合，康托展开表示的就是是当前排列组合在n个不同元素的全排列中的名次。也就是说，康托展开可以用来求一个 的任意排列的排名。

他的具体原理也不复杂，不过没啥用，不浪费时间展开说明，感兴趣百度有很多优秀的解释。

这里直接给出他的模板代码。暴力O（n^2),树状数组优化可以做到O（nlogn）.但是求排列，一般n不会很大。

所以大部分情况暴力就能解决问题。

```CPP
#include<bits/stdc++.h>
using namespace std;
using ll=long long;
ll fact[1010],P[1010],A[1010],tree[1010];//P是排列数组，fact是阶乘，自己初始化
ll lowbit(ll x)
{ 
    return x&-x; 
}
ll query(ll x)
{
    ll ans = 0;
    for (int i = x; i >= 1; i -= lowbit(i))
        ans += tree[i];
    return ans;
}
void update(ll x, ll d)
{
    for (int i = x; i <1010; i += lowbit(i))
        tree[i] += d;
}
ll cantor(int P[], int n)
{
    ll ans = 1;
    for (int i = n; i >= 1; i--)
    {
        A[i] = query(P[i]);
        update(P[i], 1);
    }
    for (int i = 1; i < n; i++)
        ans = (ans + A[i] * fact[n - i]) % MOD;
    return ans;
}
int main()
{
    cin.tie(nullptr)->sync_with_stdio(false);
    
    return 0;
}
```

### 逆康托展开

其实就是把康托展开的过程反过来，给你排名，让你求出这个排名对应的排列。

```cpp
#include<bits/stdc++.h>
using namespace std;
using ll=long long;
ll fact[MAXN] = {1}, P[MAXN],A[MAXN]; //fact需要在外部初始化
void decanter(ll x,ll n) //x为排列的排名，n为排列的长度
{
    x--;
    vector<int> rest(n, 0);
    iota(rest.begin(), rest.end(), 1); //将rest初始化为1,2,...,n
    for (int i = 1; i <= n; ++i)
    {
        A[i] = x / fact[n - i];
        x %= fact[n - i];
    }
    for (int i = 1; i <= n; ++i)
    {
        P[i] = rest[A[i]];
        rest.erase(lower_bound(rest.begin(), rest.end(), P[i]));
    }
}
```

### 卡特兰数

卡特兰数本质就是组合数学中经常出现在计数问题中的数列。他的本质其实是表达一类关系。但是在实际应用中，其实我们还是更关心他的实际价值。也就是他的值和那些问题的本质是相同的，那么我们可以通过求解卡特兰数，简介求解原问题本身。本身也是比较模板的一类问题

下面列出主要的应用场景。

1.合法路径问题，在一个网格中，从（0，0）走到（n，n）的合法方案总数。

2.你现在有n个0和n个1，问有多少个长度为2n的序列，使得序列的任意一个前缀中1的个数都大于等于0的个数，问有多少合法序列

3.你有n个左括号，n个右括号，问有多少个长度为2n的括号序列使得所有的括号都是合法的

4.有一个栈，我们有2n次操作，n次进栈，n次出栈，问有多少中合法的进出栈序列

5.一个长度为n的排列a，只要满足i<j<k且aj<ak<ai就称这个排列为312排列，求n的全排列中不是312排列的排列个数（他的本质就是1，2，。。。依次进栈，312数列就代表不可法的进出站序列）

6.在一个圆周上分布着 2n个点，两两配对，并在这两个点之间连一条弦，要求所得的2n条弦彼此不相交的配对方案数

7.有n个点，问用这n个点最终能构成多少二叉树

8.一个凸的n边形，用直线连接他的两个顶点使之分成多个三角形，每条直线不能相交，问一共有多少种划分方案

本质上这些问题的数学意义一般都和卡特兰数量的某种表达方式类似，所以才相等，实际问题中，我们也可以尝试思考，观察问题和卡特兰数列意义上的关系，或者思考新问题和上述经典模型是否相似。

实际问题中，我们求解卡特兰数的方法一般是直接组合数求解$$ F_{n}=C_{2n}^{n}−C_{2n}^{n-1} $$

### 默慈金数

这里先给出百度百科中的定义

默慈金数是在数学中，一个给定的数n的**默慈金数**是“在一个圆上的n个点间，画出彼此不相交的弦的全部方法的总数”。

他也是一个数列（OEIS中的数列A001006）表达了某种数学意义。

关于他的递推公式可以看这篇论文详细了解，[[Motzkin Number（默慈金数） - 豆丁网 (docin.com)](https://www.docin.com/p1-964777006.html)](https://www.docin.com/p1-964777006.html)

[[Motzkin Number -- from Wolfram MathWorld](https://mathworld.wolfram.com/MotzkinNumber.html)](https://mathworld.wolfram.com/MotzkinNumber.html)

这里直接给出他的递推形式$$ f_n=\frac{(2n+1)f_{n-1}+(3n-3)f_{n-2}}{n+2} $$

他还和卡特兰数有一定联系$$ f_n=\sum_{i=0}^{\lfloor{\frac{n}{2}}\rfloor}C_{n}^{2i}Cat(i) $$

从 (0,0) 到 (N,0)，你每次可以向右上一格/向右下一格/向正右一格，且不能低于y 轴的合法方案数就是默慈金数的对应值。

 本质上，上升x次就会下降x次，其他步就是向右走，所以问题本质还是从n步中选2*x步上下。剩下的部分就相当于卡特兰数的场景2，也就是这个数的定义。

除了这些直接本质相同的问题，还有一些问题会间接的用到他。

比如有一个 1∗n的矩阵，固定第一个数为1，其他填正整数， 且相邻数的差不能超过1，求方案数。

不难发现其实答案的递推数列就是$$ f_n=f_{n-1}*3-g_{n-2}$$

这里的g就是默慈金数，其实意义就是表示从 (0,0)走到 (i,0)可以向上，向下向右走一格，但是只能在第一象限的方案数。

然后问题本身就可以抽象为每一步可以有3个选择，在减去所有的非法方案即可，非法的方案其实就是前i-1步最后刚好落到1的方案，这样我们就不能继续往下走下一步。

这类问题的本质，还是要从意义考虑，一般不会直接出原始模型。

### 那罗延数

在组合数学中，那罗延数以及由那罗延数形成的那罗延三角，经常会出现在各种各样的计数问题中。

他其实也和卡特兰数有很紧密的关系。

先给出他的定义和他和卡特兰数的关系表达式。

$$ N(n,k)=\frac{1}{n}C_n^kC_n^{k-1}$$

$$ \sum_{i=1}^{n}N(n,i)=Cat(n)$$ 

实际上，他就是考虑从(0,0)到(2n,0),只有(1,1)和(-1，-1)两种移动方式，且要求运动轨迹不能低于x轴，峰为k个的合法方案数 。

 
![20210210234619668](https://user-images.githubusercontent.com/73392240/196035888-0878383d-1208-4e80-bf2e-b807e05b16d6.jpg)

所以它也可以从卡特兰数本身出发解决一些类似的问题。

比如，有 n 对括号，共有 k 对左括号和右括号相邻的合法括号序列方案数为 N(n,k)。

实际应用要灵活考虑。 

### 施罗德数

施罗德数还有一个别名，叫超级卡特兰数。

在组合数学中，施罗德数用来描述从(0,0)到(n,n)的格路中，只能使用(1,0)、(0,1)、(1,1)三种移动方式，始终位于对角线下方且不越过对角线的路径数。

不浪费时间介绍他的定义，直接给出他的递推式。

$$ S_n=\frac{(6n-3)S_{n-1}-(n-2)S_{n-2}}{n+1}$$

除了上面给出的例子，还有一些应用，但是其实很多时候，不是特别容易把他抽象出来，变成某个特殊的模型，这种问题更多还是打表解决。或者就是很明显的和他的定义中问题等价的问题。

有一个正n多边形我们要连接一些对角线，把这个多边形分成若干个区域。要求连接的对角线不能相交，每个点可以连出也可以不连出对角线。（即最终不要求所有区域均为三角形）问总方案数。

### 其他价值比较大的特殊数列

上面是从卡特兰数出发介绍的一些数列，这类数列我们往往更多还是在意他的性质本身，接下来再举一些其他数列，并且他们的相关性比较强，不过实质上也都很类似。虽然本文不展开讲解，不过还是列举出来，感兴趣的同学可以学习更多之后在深入学习。

第一类斯特林数（轮换）

第二类斯特林数（子集）

切赫数（第三类斯特林数）

贝尔数‘。

他们已有一些平凡的递推公式，但是为什么不展开介绍呢？因为他们往往伴随着生成函数一起来考，建议大家学习了生成函数和fft，ntt之后，自行前往学习查找内容。

### 小tip

很多时候，性质往往也不容易发现，与卡特兰数紧密相关的数列也不止这几项，但是在比赛中，我们完全可以通过打表求出前几项，然后查oeis来得到这些经典数列的递推式，oeis上往往会给出更详细的数列描述，方便辨别是否是你想要的数列。并且这种方法其实比上面讲的加起来还要更实用。这种特殊数列和特殊性质我们往往还是很难考积累全部记全，出题人随便在oeis上找一个数列，选择一个意义然后抽象化另一个题面，就能出一道新题，所以更关键的还是熟练掌握打表技术。

### 自适应辛普森（Simpson）积分

很多时候，我们会面临一些求积分的问题，无论是直接给你函数，让你想办法求解积分，还是对于一些计算几何问题，无法直接推导积分，我们都可以用这种方法来求一段区域的积分，积分的相关基础概念这里就不再赘述，今天主要就是说明他的大致原理，和他的用法。

他的本质就是把函数看作一个二次函数，如果区间够小，那么函数就足够近似，我们可以直接按照二次函数来计算每个小段的积分。

推导其实就是最基本的积分过程

​$$ \int_{a}^{b} f(x) dx=\int_{a}^{b}(Ax^2+Bx+C)dx$$

$$ =\frac{A}{3}(b^3-a^3)+\frac{B}{2}(b^2-a^2)+C(b-a)$$

$$ =\frac{2A(b^3-a^3)+3B(b^2-a^2)+6C(b-a)}{6}$$

$$ =\frac{(b-a)(Aa^2+Ba+C+Ab^2+Bb+C+Aa^2+2Aab+Ab^2+2Bb+2Ba+4C)}{6}$$

$$ =\frac{(b-a)(f(a)+f(b)+A(a+b)^2+2B(a+b)+4C)}{6}$$

$$ =\frac{(b-a)(f(a)+f(b)+4(A(\frac{(a+b)}{2})^2+B(\frac{a+b}{2})+C)}{6}$$

$$ =\frac{(b-a)(f(a)+f(b)+4f(\frac{a+b}{2})}{6}$$

于是我们就得到了一个近似的辛普森公式。

b-a越小，他就越接近正确答案。

一个普通的思路就是，把要求的积分区间分解成一个个小的区间，然后累加在一起，但是这样做的缺点就是，如果为了保证精度，每个区间都会很小，所以这样就很容易会TLE。那么如何控制这个区间范围比较好呢？

其实就是把问题不断分解，比如求一个区间（l，r）的积分。

先取一个mid=（l+r）/2，在对两部分分别计算辛普森积分，并且对整个区间计算辛普森积分，然后判断他们的误差是否在精度要求内，如果在就直接返回，否则继续递归下去，很朴素的一个分治思想。代码也很短。

```CPP
#include<bits/stdc++.h>
using namespace std;
using ll=long long;
double a,b,c,d,l,r;
const double eps=1e-8;
double f(double x)//原函数
{  
	return  ;       
}
double simpson(double l,double r) //Simpson公式
{ 
	double mid=(l+r)/2;
	return (f(l)+4*f(mid)+f(r))*(r-l)/6;
}
double solve(double l,double r,double ans) 
{
	double mid=(l+r)/2;
	double ll=simpson(l,mid),rr=simpson(mid,r);
	if(fabs(ll+rr-ans)<=eps) return ll+rr;     //确认精度
	return solve(l,mid,ll)+solve(mid,r,rr);     //精度不够则递归调用
}
int main()
{
    cin.tie(nullptr)->sync_with_stdio(false);
    //cin>>a>>b>>c>>d>>l>>r;
    cout<<fixed<<setprecision(6)<<solve(l,r,simpson(l,r))<<"\n";
    return 0;
}
```

不过需要注意的是，在实际应用是，要灵活调整精度，否则可能在wa和tle之间反复徘徊。

这里在介绍一个优化技巧。

```CPP
double solve(double l,double r,double eps,double ans) 
{
	double mid=(l+r)/2;
	double ll=simpson(l,mid),rr=simpson(mid,r);
	if(fabs(ll+rr-ans)<=15*eps) return ll+rr+(ll+rr-ans)/15.0;     //确认精度
	return solve(l,mid,eps/2,ll)+solve(mid,r,eps/2,rr);     //精度不够则递归调用
}
```

这样对于每个区间，我都可以保证，整个区间的精度可以继续保持。

至于为啥要弄15， 这里有个论文给了解释，感兴趣可以了解，不感兴趣，存个板子直接用就好了。 

 
[AdaptiveQuadProof.pdf](https://github.com/CSGrandeur/s-1problem1day1ac/files/9806343/AdaptiveQuadProof.pdf)

**Comment by CSGrandeur** [comment]
部分特殊计数序列题： [soj-特殊计数序列](https://soj.csgrandeur.cn/csgoj/problemset#search=%E7%89%B9%E6%AE%8A%E8%AE%A1%E6%95%B0%E5%BA%8F%E5%88%97)

参考：《组合数学（原书第5版）》
